src/chromium/crBrowser.ts

```diff
@@ -105,87 +107,103 @@ export class CRBrowser extends platform.EventEmitter implements Browser {
+        if (opener && opener._initializedPage && !opener._initializedPage.isClosed())
+          opener._initializedPage.emit(CommonEvents.Page.Popup, crPage._page);
       });
       return;
     }
-    assert(targetInfo.type === 'service_worker');
-    target.serviceWorkerOrError().then(workerOrError => {
-      if (workerOrError instanceof Worker)
-        context.emit(Events.CRBrowserContext.ServiceWorker, workerOrError);
-    });
-  }
 
-  private _createTarget(targetInfo: Protocol.Target.TargetInfo, session: CRSession) {
-    const {browserContextId} = targetInfo;
-    const context = (browserContextId && this._contexts.has(browserContextId)) ? this._contexts.get(browserContextId)! : this._defaultContext;
-    let hasInitialAboutBlank = false;
-    if (CRTarget.isPageType(targetInfo.type) && targetInfo.openerId) {
-      const openerIndex = this._popupOpeners.indexOf(targetInfo.openerId);
-      if (openerIndex !== -1) {
-        this._popupOpeners.splice(openerIndex, 1);
-        // When this page is a result of window.open($url) call, we should have it's opener
-        // in the list of popup openers. In this case we know there is an initial
-        // about:blank navigation, followed by a navigation to $url.
-        hasInitialAboutBlank = true;
-      }
+    if (targetInfo.type === 'service_worker') {
+      const serviceWorker = new CRServiceWorker(context, targetInfo.url);
+      this._serviceWorkers.set(targetInfo.targetId, serviceWorker);
+      this._initializeServiceWorker(serviceWorker, context, session);
+      return;
+    }
+
+    assert(targetInfo.type === 'browser' || targetInfo.type === 'other');
+    if (waitingForDebugger) {
+      // Ideally, detaching should resume any target, but there is a bug in the backend.
+      session.send('Runtime.runIfWaitingForDebugger').catch(debugError).then(() => {
+        this._session.send('Target.detachFromTarget', { sessionId }).catch(debugError);
+      });
     }
-    const target = new CRTarget(this, targetInfo, context, session, hasInitialAboutBlank);
-    assert(!this._targets.has(targetInfo.targetId), 'Target should not exist before targetCreated');
-    this._targets.set(targetInfo.targetId, target);
-    return { context, target };
   }
 
-  _onDetachedFromTarget({targetId}: Protocol.Target.detachFromTargetParameters) {
-    const target = this._targets.get(targetId!)!;
-    if (!target)
+  _initializeServiceWorker(serviceWorker: Worker, context: CRBrowserContext, session: CRSession) {
+    session.once('Runtime.executionContextCreated', event => {
+      serviceWorker._createExecutionContext(new CRExecutionContext(session, event.context));
+    });
+    // This might fail if the target is closed before we receive all execution contexts.
+    session.send('Runtime.enable', {}).catch(e => {});
+    session.send('Runtime.runIfWaitingForDebugger').catch(e => {});
+    context.emit(Events.CRBrowserContext.ServiceWorker, serviceWorker);
```

Comment:
Let's put session initialization logic into a method on `CRServiceWorker` and move `context.emit(Events.CRBrowserContext.ServiceWorker` to `_onAttachedToTarget` so that all notifications are fired in one method.